### PR TITLE
refactor: migrate to rules_oci

### PR DESCRIPTION
**Pull request by thesayyn**
Remaining work here; 

- [x] Update root `BUILD` file to use rules_oci  `oci_index` `oci_push` `oci_cosign`
- [x] http_archive_sbom implementation is not complete yet.
- [x] Update cloudbuild to use bazel rules for signing and publishing
- [x] Remove `cloudbuild_*.sh` scripts as not needed anymore
- [x] Some cleanup work

### CODE REVIEW

File: base/BUILD [hunk 1]
```diff
@@ -1,11 +1,10 @@
-package(default_visibility = ["//visibility:public"])
-
 load(":base.bzl", "NONROOT", "distro_components")
 load(":distro.bzl", "DISTROS")
 load("@bazel_tools//tools/build_defs/pkg:pkg.bzl", "pkg_tar")
```
Comment (loosebazooka) [inline base/BUILD:3]: should this use: load("@rules_pkg//:pkg.bzl", "pkg_tar") ?

Comment (thesayyn) [reply base/BUILD]: thanks. fixed it!

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -8,59 +8,55 @@ on:
...
+            path: |
+                ~/.cache/bazel
+                ~/.cache/bazel-repo
+            key: bazel-cache-image-${{ hashFiles('**/BUILD.bazel', '**/*.bzl', '**/*.yaml', 'WORKSPACE') }}-${{ github.sha }}
+            restore-keys: |
+                bazel-cache-image-
+                bazel-cache-
+      - name: Free space
+        run: |
+          sudo apt-get remove -y '^ghc-8.*'
```
Comment (loosebazooka) [inline .github/workflows/ci.yaml:29]: looks like this is causing some failures since the packages isn't there?

Comment (thesayyn) [reply .github/workflows/ci.yaml]: fixed now.

File: WORKSPACE [hunk 3]
```diff
@@ -1,6 +1,35 @@
...
+load("@contrib_rules_oci//oci:dependencies.bzl", "rules_oci_dependencies")
+
+rules_oci_dependencies()
+
+load("@contrib_rules_oci//oci:repositories.bzl", "LATEST_CRANE_VERSION", "LATEST_ZOT_VERSION", "oci_register_toolchains")
+
+oci_register_toolchains(
+    name = "oci",
+    crane_version = LATEST_CRANE_VERSION,
+    zot_version = LATEST_ZOT_VERSION,
```
Comment (loosebazooka) [inline WORKSPACE:27]: where are these values coming from?

File: debian_package_manager/internal/rhttp/rhttp.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+do:
+	resp, err := _http.Get(url)
+	if err == nil {
+		if resp.StatusCode == _http.StatusGatewayTimeout && retry < 10 {
+			fmt.Printf("Retrying: %s\n", url)
+			retry += 1
+			time.Sleep(time.Second * time.Duration(retry))
+			goto do
+		}
+		fmt.Printf("URL: %s, Status: %s\n", url, resp.Status)
```
Comment (loosebazooka) [inline debian_package_manager/internal/rhttp/rhttp.go:20]: I think this line should be after 13 for better readability if a retry occurs.

File: nodejs/control [hunk 5]
```diff
@@ -0,0 +1,16 @@
+Package: nodejs
```
Comment (loosebazooka) [inline nodejs/control:1]: How does this part work? Is this information necessary for sbom creation? I'm worried about the "maintainer" aspect.

File: private/oci/sign_and_push.bzl [hunk 6]
```diff
@@ -0,0 +1,98 @@
+load("@contrib_rules_oci//oci:defs.bzl", "oci_push")
+load("@contrib_rules_oci//cosign:defs.bzl", "cosign_attach", "cosign_sign")
+load("//private/pkg:oci_image_spdx.bzl", "oci_image_spdx")
+
+PUSH_AND_SIGN_CMD = """\
+repository="$(stamp "{REPOSITORY}")"
+tag="$(stamp "{TAG}")"
+[[ -n $EXPORT ]] && echo "$repository:$tag" >> $EXPORT
+"$(realpath {ATTACH_CMD})" --repository "$repository"
+"$(realpath {SIGN_CMD})" --repository "$repository" --key "$KEY" --attachment sbom
```
Comment (loosebazooka) [inline private/oci/sign_and_push.bzl:10]: Can we add keyless signing here too? I vaguely recall saying I didn't think people used it, but I think they actually do.

File: private/pkg/oci_image_spdx.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
+package main
```
Comment (loosebazooka) [inline private/pkg/oci_image_spdx.go:1]: can a simple test be added for this? Would make it clearer what the expected output is.

**Review by loosebazooka — COMMENTED** [review]
looks good, I would like to run it though our "/testing" repo that we did for caching to make sure everything works.

File: WORKSPACE (hunk 3 above)
Comment (thesayyn) [reply WORKSPACE:27]: those come from rules_oci

File: nodejs/control (hunk 5 above)
Comment (thesayyn) [reply nodejs/control:1]: we could change the maintainer fields. I took it from nodesources deb to have similar control files. this file is used by both sbom and dpkg. sbom converts this file to spdx and dpkg just puts it into its respective location.

File: private/oci/sign_and_push.bzl (hunk 6 above)
Comment (thesayyn) [reply private/oci/sign_and_push.bzl:10]: by keyless, you mean using a KMS? if yes, that's the case here. it uses gcp's km same as the cloudbuild_cosign script did.

File: private/pkg/oci_image_spdx.go (hunk 7 above)
Comment (thesayyn) [reply private/pkg/oci_image_spdx.go:1]: I'll do that. nice suggestion.

File: private/oci/sign_and_push.bzl (hunk 6 above)
Comment (loosebazooka) [reply private/oci/sign_and_push.bzl:10]: sorry no I mean whats in: https://github.com/GoogleContainerTools/distroless/blob/main/.cloudbuild/cloudbuild.yaml#L46-L59

Comment (loosebazooka) [reply private/oci/sign_and_push.bzl:10]: cosign's signing using identity rather than keys.

File: nodejs/control (hunk 5 above)
Comment (loosebazooka) [reply nodejs/control:1]: I might just point it to nodejs.org instead of distroless

**Comment by loosebazooka** [comment]
is the github cache running out of space?

File: private/oci/sign_and_push.bzl (hunk 6 above)
Comment (thesayyn) [reply private/oci/sign_and_push.bzl:10]: sure. i'll do that.

File: nodejs/control (hunk 5 above)
Comment (thesayyn) [reply nodejs/control:1]: i can commit the changes if you could provide a diff here.

**Comment by thesayyn** [comment]
> is the github cache running out of space?

my guess too. though I can't tell for sure since logs are absent.

File: nodejs/control (hunk 5 above)
Comment (thesayyn) [reply nodejs/control:1]: friendly ping

File: debian_package_manager/internal/rhttp/rhttp.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	"time"
+)
+
+func Get(url string) (*_http.Response, error) {
+	retry := 0
+do:
+	resp, err := _http.Get(url)
+	if err == nil {
+		if resp.StatusCode == _http.StatusGatewayTimeout && retry < 10 {
+			fmt.Printf("URL: %s, Status: %s\n", url, resp.Status)
```
Comment (loosebazooka) [inline debian_package_manager/internal/rhttp/rhttp.go:15]:
```suggestion
	if err == nil {
		fmt.Printf("URL: %s, Status: %s\n", url, resp.Status)
		if resp.StatusCode == _http.StatusGatewayTimeout && retry < 10 {
```

File: nodejs/control [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+Version: {{VERSION}}
+Architecture: {{ARCHITECTURE}}
+Maintainer: Distroless <https://github.com/GoogleContainerTools/distroless>
```
Comment (loosebazooka) [inline nodejs/control:4]:
```suggestion
Maintainer: OpenJS Foundation <https://openjsf.org/>
```

File: .cloudbuild/cloudbuild.yaml [hunk 10]
```diff
@@ -17,43 +17,6 @@ steps:
       - REGISTRY=gcr.io
```
Comment (loosebazooka) [inline .cloudbuild/cloudbuild.yaml:17]: PROJECT_ID -> PROJECT_ID/testing for now?

**Review by loosebazooka — COMMENTED** [review]
looks great. just want to do it into a testing branch agian.

**Comment by thesayyn** [comment]
done.

**Pull request merged**